lib/azure/armrest/configuration.rb

```diff
@@ -0,0 +1,256 @@
+      #
+      def token
+        @token, @token_expiration = tokens[cache_key] if @token.nil?
+
+        if @token.nil? || Time.now.utc > @token_expiration
+          @token, @token_expiration = fetch_token
+        end
+
+        @token
+      end
+
+      # Set the token value and expiration time.
+      #
+      def set_token(token, token_expiration)
+        validate_token_time(token_expiration)
+
+        tokens[token].delete_if { |time| Time.now.utc > time.utc }
+        tokens[token] << token_expiration
+
+        @token, @token_expiration = token, token_expiration
+      end
+
+      # Returns the expiration datetime for given key, or the current
+      # cache_key if no key is specified.
+      #
+      def token_expiration(key = cache_key)
+        tokens[key].last
+      end
+
+      # Set the time in which the token expires. The time is automatically
+      # converted to UTC.
+      #
+      def token_expiration=(time)
+        @token_expiration = time.utc
+      end
+
+      # The name of the file or handle used to log http requests.
+      #--
+      # We have to do a little extra work here to convert a possible
+      # file handle to a file name.
+      #
+      def self.log
+        file = RestClient.log.instance_variable_get("@target_file")
+        file || RestClient.log
+      end
+
+      # Sets the log to +output+, which can be a file or a handle.
+      #
+      def self.log=(output)
+        RestClient.log = output
+      end
+
+      private
+
+      # A list of tokens used with the current set of credentials.
+      #
+      def tokens
+        self.class.token_cache
+      end
+
+      # A combination of grant_type, tenant_id, client_id and client_key,
+      # as a single string joined by underscores.
+      #
+      # Used to identify unique sessions.
+      #
+      def cache_key
+        [tenant_id, client_id, client_key].join('_')
+      end
+
+      # Don't allow tokens from the past to be set.
+      #
+      def validate_token_time(time)
+        if time.utc < Time.now.utc
+          raise ArgumentError, 'token_expiration date invalid'
+        end
+      end
+
+      # Build a one-time lookup hash that sets the appropriate api-version
+      # string for a given provider and resource type. If possible, select
+      # a non-preview version that is not set in the future. Otherwise, just
+      # just the most recent one.
+      #
+      def set_provider_info(providers)
+        providers.each do |rp|
+          rp.resource_types.each do |rt|
+            if rt.api_versions.any? { |v| v !~ /preview/i && Time.parse(v).utc <= Time.now.utc }
+              api_version = rt.api_versions.reject do |version|
+                version =~ /preview/i || Time.parse(version).utc > Time.now.utc
+              end.first
+            else
+              api_version = rt.api_versions.first
+            end
+
+            self.class.provider_cache[rp.namespace.downcase].merge!(rt.resource_type.downcase => api_version)
```

Comment:
We used to have `namespace` naming conflict, can you double check it is no longer a problem running at miq env.

Since this cache is used solely for api_version look up, you may want to rename `provider_cache` to something like `api_version_cache` or `api_version_lookup`